### ISSUE TITLE
Lowering failure for mhlo.pad of empty tensor

### ISSUE THREAD

**Issue by hawkinsp** [openxla/iree#9297]
```
import iree.compiler

CODE = """
#loc0 = loc(unknown)
module @jit_prim_fun.0 {
  func.func public @main(%arg0: tensor<0x2xi16> loc(unknown), %arg1: tensor<i16> loc(unknown)) -> tensor<3x3xi16> {
    %0 = "mhlo.pad"(%arg0, %arg1) {edge_padding_high = dense<[2, 1]> : tensor<2xi64>, edge_padding_low = dense<[1, 0]> : tensor<2xi64>, interior_padding = dense<[1, 0]> : tensor<2xi64>} : (tensor<0x2xi16>, tensor<i16>) -> tensor<3x3xi16> loc(#loc1)
    return %0 : tensor<3x3xi16> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)
#loc1 = loc("jit(pad)/jit(main)/pad[padding_config=((1, 2, 1), (0, 1, 0))]"("/Users/phawkins/p/jax/tests/lax_test.py":1419:1))
"""

iree_binary = iree.compiler.compile_str(
    CODE, target_backends=["dylib"], input_type="mhlo")

```

```
Traceback (most recent call last):
  File "/Users/phawkins/p/jax/t.py", line 14, in <module>
    iree_binary = iree.compiler.compile_str(
  File "/Users/phawkins/.pyenv/versions/py310/lib/python3.10/site-packages/iree/compiler/tools/core.py", line 293, in compile_str
    result = invoke_immediate(cl, immediate_input=input_bytes)
  File "/Users/phawkins/.pyenv/versions/py310/lib/python3.10/site-packages/iree/compiler/tools/binaries.py", line 200, in invoke_immediate
    raise CompilerToolError(process)
iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool iree-compile
Diagnostics:
/Users/phawkins/p/jax/tests/lax_test.py:1419:1: error: expected result type to be 'memref<0x2xi16, affine_map<(d0, d1) -> (0)>>' or a rank-reduced version. (mismatch of result layout)
    fun = lambda operand: lax.pad(operand, np.array(0, dtype), pads)
^
/Users/phawkins/p/jax/tests/lax_test.py:1419:1: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"llvm", "embedded-elf-x86_64", {cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
    fun = lambda operand: lax.pad(operand, np.array(0, dtype), pads)
^
/Users/phawkins/p/jax/tests/lax_test.py:1419:1: error: failed to serialize executables
    fun = lambda operand: lax.pad(operand, np.array(0, dtype), pads)
^
compilation failed


Invoked with:
 iree-compile /Users/phawkins/.pyenv/versions/py310/lib/python3.10/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=dylib --iree-mlir-to-vm-bytecode-module --iree-llvm-embedded-linker-path=/Users/phawkins/.pyenv/versions/py310/lib/python3.10/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false

```

**Comment by benvanik** [comment]
This is likely a duplicate of #9293 (memref with a 0 dimension).

**Comment by MaheshRavishankar** [comment]
Dont think #9293 is related. 

This has a `tensor<0x2xi16>` . THis is invalid shape. I dont think we support such cases. Might be Python -> MHLO conversion error?

**Comment by MaheshRavishankar** [comment]
This has a tensor<0x2xi16> . THis is invalid shape. I dont think we support such cases. Might be Python -> MHLO conversion error?

**Comment by hawkinsp** [comment]
Zero-sized shapes are legal according to the only HLO spec we have, namely XLA accepts them. I admit that is partially because I made XLA accept them years ago. We should make this explicit in the MHLO spec (either way).

My personal belief is that zero-sized shapes are very useful as a base-case; it is considerably more inconvenient to work around their absence at the Python level than it is to add a compiler pass that removes them. The latter is what XLA, being a statically shaped compiler, does: it has an [early HLO pass](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/zero_sized_hlo_elimination.h;l=22) that replaces all zero-sized operators with constants and then the need to handle zero-element shapes is limited to only a few places.

I have less opinion about dynamically-zero shapes: I can believe they may be significant harder to deal with.

**Comment by benvanik** [comment]
Agreed on supporting 0 dims; they should be trivial to statically handle at the mhlo/linalg level via propagation, and then at runtime most things should just work as the workgroup counts we calculate based on zeros should (if we're doing it right) end up as zeros as well. We can make not dispatching empty work an optimization but in the meantime still be correct.

**Comment by sdasgup3** [comment]
>> We should make this explicit in the MHLO spec (either way).
I will work on this.

**Comment by MaheshRavishankar** [comment]
I dont fully understand what the semantics of the 0-dim tensor would be. It leads to some very weird behavior. For example

```
%1 = tensor.extract_slice %0[%a, %b] [1, %c] [1, 1] : tensor<0x?xf32> to tensor<?xf32>
```

is creating a real tensor out of an empty tensor. I'd like to head off such `tensor<...x0x...>` before it even gets to Linalg, so during MHLO to Linalg conversion itself. The question for me though is if you have a `tensor<..x0x...>` what do you replace it with (especially if the other dimensions are dynamic). I cant wrap my head around making some code dead when the value still has uses.

**Comment by hawkinsp** [comment]
Why would that `tensor.extract_slice` be legal? Extracting a size 1 slice from a size 0 dimension should be an error, just as extracting a size 10 slice from a size 9 dimension would be.

**Comment by stellaraccident** [comment]
> is creating a real tensor out of an empty tensor. I'd like to head off such tensor<...x0x...> before it even gets to Linalg, so during MHLO to Linalg conversion itself. The question for me though is if you have a tensor<..x0x...> what do you replace it with (especially if the other dimensions are dynamic). I cant wrap my head around making some code dead when the value still has uses.

I was thinking of an MHLO level pass which eliminated the 0-dim ops up front (replacing with a constant). I haven't worked through all of the details, but Peter's prior work on this seems a strong prior.

**Comment by MaheshRavishankar** [comment]
> Why would that `tensor.extract_slice` be legal? Extracting a size 1 slice from a size 0 dimension should be an error, just as extracting a size 10 slice from a size 9 dimension would be.

Good point on that. Lets leave aside the size being 1. Lets say the operation is
```
%1 = tensor.extract_slice %0[%a, %b] [0, %c] [1, 1] : tensor<0x?xf32> to tensor<?xf32>
```
`tensor.extract_slice` allow rank-reducing semantics. So if a size of the slice is `1`, you can drop that dimension. (It doesnt really work when the size is 0, nor does it check the size is not 0 cause that is all undefined/unhandled). So I was just kind of using the rank-reducing semantics. We could disallow the size being 0, but that only works for static cases. Typically the fact that the size is 0 is only known after some canonicalizations are run. So it will end up failing in a canonicalizer which is very unpleasant.

@hawkinsp what constant value did you replace the tensor with 0-dims with? I have a heard time seeing how these can be made to connect, so some pointers would help.

**Comment by hawkinsp** [comment]
The idea in the XLA rewrite was: ok, any array with zero elements *is* constant. So you get to replace almost everything that produces or consumes a size 0 array with such a constant. You must still handle 0-element constants. But not really much else in the way of other ops! Again, this was in the context of a statically shaped compiler...

Note there *are* cases that aren't completely trivial, e.g., contractions or reductions over size 0 dimensions. But they are still things you can constant fold if you have static shapes.

**Comment by MaheshRavishankar** [comment]
Its a bit tricky within MLIR. The SSA value that is the `tensor<...x0x...>` is not dead. It has other uses. It could be used as an operand in an operation that does not itself return a `tensor<...x0x...>` (similar to the contractions/reductions example you mentioned). You need to delete all uses of the operation before you delete an operation. You basically need to get the "frontier of influence" of any value that is `tensor<...x0x...>` and start removing those. I am seeing any operation involving a `tensor<...x0x...>` as dead code (or an operation that is really just there to propagate metadata) but I cant find the point at which I can start removing these operations.

**Comment by MaheshRavishankar** [comment]
> Its a bit tricky within MLIR. The SSA value that is the `tensor<...x0x...>` is not dead. It has other uses. It could be used as an operand in an operation that does not itself return a `tensor<...x0x...>` (similar to the contractions/reductions example you mentioned). You need to delete all uses of the operation before you delete an operation. You basically need to get the "frontier of influence" of any value that is `tensor<...x0x...>` and start removing those. I am seeing any operation involving a `tensor<...x0x...>` as dead code (or an operation that is really just there to propagate metadata) but I cant find the point at which I can start removing these operations.

Actually I re-read your comment and I missed something. Now I understand what you are suggesting... 

THis will also need the uses in individual operations resolved. For example, we would need to modify a `tensor.pad` that takes `tensor<...x0x...>` operand into something that captures the same semantics but with a different type.

**Comment by stellaraccident** [comment]
Mahesh: what if we just rauw any operation that we determine is a 0dim tensor? I think if we just do that to *everything* we get to avoid a dominant frontier analysis: they just all become a constant.

**Comment by MaheshRavishankar** [comment]
I dont know if that is enough. Take this example itself. It is taking a `tensor<...x0x...>` and create a new tensor that has all non-zero extents. So if you treat the operand as a constant value, you are now creating a new tensor with a legitimate value. So the computation is not dead, though my instinct tells me that a 0-dim tensor is basically saying its dead. The output of the computation is now heavily dependent on the arbitrary constant value we chose to replace the operand with. Definitely thats not the programmer intent.

I'd really like to know what exactly is the semantics of a tensor with any extent being 0. I really cannot wrap my head around anything other than it being dead code, but the way it is used makes it really hard to remove all of these.

**Comment by stellaraccident** [comment]
<hand-waving>
I suspect the definition of when the result can be considered to be 0-extend is op specific (as peter alluded to, there are cases that are non-trivial) but can be defined -- at least with respect to static 0-dims. I think that the pad in that case might just become a fill of a constant value, no?
</hand-waving>

It would be really handy if MHLO actually had a spec. This is the type of thing that just a bit of formal definitions would clean up, imo.

**Comment by benvanik** [comment]
Yeah, pads should fill the padded elements (padded_dim - original_dim) and if original_dim=0 and padded_dim>0 that's totally fine - it just means it's a fill.

`flow.tensor.empty` already deals with this using RAUW - we don't replace tensors with constants but with a specific value that carries all the metadata but never allows for materialization. A linalg.empty_tensor (ala linalg.init_tensor) could do something similar while still capturing all dimension metadata. It's always going to be valid to move these 0 element tensors around - we want to support things like `select %cond, %actual_value, %empty_value` or branch/function arguments that are empty (start with empty tensor and then (re)initialize it in a loop body, etc), or default global values of dynamically shaped tensors are going to start as empty, etc. That is to say, we're never going to be able to fully get rid of them and have to treat the entire RAUW process only as an optimization - but an important one.

**Comment by MaheshRavishankar** [comment]
> `<hand-waving>` I suspect the definition of when the result can be considered to be 0-extend is op specific (as peter alluded to, there are cases that are non-trivial) but can be defined -- at least with respect to static 0-dims. I think that the pad in that case might just become a fill of a constant value, no? `</hand-waving>`
> 

I suspect this is the case indeed. So we need a pass that inspects every operation with an operand that has a tensor of zero-dim and handle it appropriately. Then we will be implementing "a spec" on the fly, so not sure if it signifies programmer intent. If the op-specific handling we implement leads to a difference in value compared to what XLA does (when we compare with XLA for correctness) root causing it to this will be extremely hard.

**Comment by stellaraccident** [comment]
To be clear, I think that such a pass lives in MHLO and they need to own coherence of the details. We can seed that, but shouldn't own it. We will need to own any of the results of having attempted to maximally redact 0-extents.

**Comment by MaheshRavishankar** [comment]
That make sense. FYI @jpienaar maybe something to sync with MHLO team.

Replacing all operands that have `tensor<...x0x..>` with a `linalg.init_tensor` of the same shape and then using shape propagation to truly ensure that all the metadata has been resolved is the best we can do. If there are any `linalg.init_tensor` remaining those get converted to `flow.tensor.empty` (I think that happens already). 

(Apologies if this was already the state of the world everyone here saw. I took me a while to get there).

**Comment by stellaraccident** [comment]
No apology necessary - it took mea minute as well, and I had the benefit of having been on some of these threads years ago with TF.

Just to check assumptions: is this necessarily an HLO level rewrite or can it be done once in linalg? It seems like it might just be a few patterns at the linalg level, and we are better setup for metadata-only tensors there (and would have a good indication of when the transformation may be incomplete in some form)?

**Comment by jpienaar** [comment]
What you/Mahesh proposes sounds not too tricky during MHLO to LinAlg lowering (probability dedicated pass "cleanup" pass just for this pattern given the ordering of lowering to LinAlg or LinAlgExt or what not differs). I don't know how invasive it would be linalg side (I mean the linalg side should work with dynamic dims where one would encounter this, so treating a zero same as an unknown could work out of box, not saying way to actually implement).

**Comment by benvanik** [comment]
Though MHLO->linalg would be nice (the less IR generated and that survives through lowering the better!), we still need to support this on TOSA/torch paths as well as for shapes that end up as 0 after we've lowered into linalg and done consteval and IPO - quirks could be handled at the higher levels but we're going to need to support this generically in linalg unless we want a whackamole nightmare.

**Comment by jpienaar** [comment]
SG, MHLO lowering path should be very straightforward  and then we can unblock here while LinAlg is ongoing.

**Comment by MaheshRavishankar** [comment]
Lets split the use cases into two. That will help me state my reasoning more clearly.

1) The `tensor<...x0x...>` is only there for carrying metadata. In other words the union of forward slices of all uses of such tensors are there for the sole purpose of carrying shape information and eventually the only useful computation out of this union of slices is `tensor.dim` operations. I would claim that is already supported today (even without replacing all uses with `linalg.init_tensor`)
2) Uses of `tensor<...x0x...>` in operations requires op specific reasoning to resolve. For example,
    a) When used as a source of a pad operation, it can be replaced by a fill of the padded value of the result shape
    b) As a consequence of (a), such tensors used as source of a `tensor.insert_slice` is a no-op and the result can be replaced with its `dest`.
    c) When used as operands of a Structured op, if the 0-dim results in a 0-trip loop, then the results of the Structured op can be replaced with `linalg.init_tensor`. 
All these individually make sense, but I dont see a reason to codify this behavior. As far as I can see all cases where such op-specific reasoning is needed to remove such `tensor<...x0x...>` are all symptoms of a either a wrong program or a wrong lowering. It would be strange to have an actual use case where the expected output is "undef". There might be unit tests where the expected result is undef, but those are contrived in my opinion. I'd rather flag these as errors instead of compiling and returning "undef". I dont see how that serves a legitimate use case (apart from some unit tests passing). If such semantics are "supported" by MHLO, I'd rather limit the solution to MHLO instead of making it a general supported feature, there by setting this expectation in other front ends.

**Comment by jpienaar** [comment]
Looking at top level issues, moved textual MLIR to input.mlir and tried to run. Luckily the textual format had not changed since the issue was filed. With input.mlir as

```mlir
module @jit_prim_fun.0 {
  func.func public @main(%arg0: tensor<0x2xi16>, %arg1: tensor<i16>) -> tensor<3x3xi16> {
    %0 = "mhlo.pad"(%arg0, %arg1) {edge_padding_high = dense<[2, 1]> : tensor<2xi64>, edge_padding_low = dense<[1, 0]> : tensor<2xi64>, interior_padding = dense<[1, 0]> : tensor<2xi64>} : (tensor<0x2xi16>, tensor<i16>) -> tensor<3x3xi16>
    return %0 : tensor<3x3xi16>
  }
}
```

and running iree-compile from build directory:

```sh
build/tools/iree-compile --iree-hal-target-backends=llvm-cpu -iree-input-type=mhlo /tmp/input.mlir -o /dev/null
```

passes without any failure. So seems to be all good now for the original case. I don't know if any of the other issues here are pending still, but closing given issue addressed and we can file new issues if needed.
